// playwright-viewer/lib/viewer.js
// lib/viewer.js
// Generic viewer server for Playwright custom JSON reports.
// Exposes: createServer({ reportDir?, port? }) -> http.Server

const http = require('http');
const fs = require('fs');
const path = require('path');
const url = require('url');

function createServer(options = {}) {
  const reportRoot = options.reportDir || path.join(process.cwd(), 'custom-report');
  const port = options.port || parseInt(process.env.PORT || '4173', 10);

  const viewerHtml = `<!doctype html>
<html lang="en">
<head>
  <meta charset="UTF-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Playwright Custom Report</title>
  <style>
    :root {
      --bg: #f3f7ff;
      --bg-panel: #e1f0ff;
      --card-bg: #ffffff;
      --card-border: #d7e4f6;
      --text-main: #243b53;
      --text-muted: #7b8794;
      --accent: #4c6fff;
      --accent-soft: #edf2ff;
      --shadow-soft: 0 18px 40px rgba(15, 23, 42, 0.12);
    }
    * { box-sizing: border-box; }
    body {
      margin: 0;
      font-family: system-ui, -apple-system, Segoe UI, sans-serif;
      background: radial-gradient(circle at top left, #e6f3ff 0, var(--bg) 40%, #ffffff 100%);
      color: var(--text-main);
    }
    header {
      display: flex;
      justify-content: space-between;
      align-items: center;
      padding: 16px 24px;
      border-bottom: 1px solid #d0e2ff;
      background: rgba(255, 255, 255, 0.9);
      backdrop-filter: blur(10px);
      position: sticky;
      top: 0;
      z-index: 1;
    }
    h1 { margin: 0; font-size: 22px; font-weight: 700; }
    main {
      padding: 24px;
      max-width: 1400px;
      margin: 0 auto;
      background: linear-gradient(135deg, var(--bg-panel) 0, #f9fbff 60%, #ffffff 100%);
      border-radius: 24px 24px 0 0;
      box-shadow: 0 18px 60px rgba(15, 23, 42, 0.16);
      margin-top: 16px;
    }
    .controls { display: flex; gap: 12px; align-items: center; flex-wrap: wrap; }
    select, button {
      background: var(--accent-soft);
      color: var(--text-main);
      border: 1px solid #c3d4ff;
      border-radius: 999px;
      padding: 8px 14px;
      cursor: pointer;
      font-size: 13px;
    }
    select:focus, button:focus {
      outline: 2px solid var(--accent);
      outline-offset: 1px;
    }
    button:hover {
      background: #d9e2ff;
    }
    .grid { display: grid; gap: 16px; grid-template-columns: repeat(auto-fit, minmax(260px, 1fr)); }
    .card {
      background: var(--card-bg);
      border: 1px solid var(--card-border);
      border-radius: 18px;
      padding: 16px 18px;
      box-shadow: var(--shadow-soft);
    }
    .card h3 { margin: 0 0 8px; font-size: 15px; color: var(--text-main); font-weight: 600; }
    .metric { font-size: 32px; font-weight: 700; }
    .muted { color: var(--text-muted); font-size: 13px; }
    .badge {
      display: inline-flex;
      align-items: center;
      gap: 6px;
      padding: 3px 9px;
      border-radius: 999px;
      font-size: 12px;
      font-weight: 600;
      border: 1px solid transparent;
    }
    .pass { background: #e3f9e5; color: #037971; border-color: #c1eac5; }
    .fail { background: #ffeff0; color: #c81e1e; border-color: #fcc2c3; }
    .skip { background: #fff7e8; color: #ad6200; border-color: #ffe0a3; }
    .warn { background: #fff4e6; color: #b44d12; border-color: #ffd0a8; }
    table { 
      width: 100%; 
      border-collapse: collapse;
      table-layout: fixed;
    }
    th, td { 
      padding: 12px 8px; 
      text-align: left; 
      border-bottom: 1px solid #e2e8f0;
      height: 52px;
      vertical-align: middle;
      overflow: hidden;
    }
    th { color: #4c6fff; font-weight: 600; font-size: 13px; height: auto; }
    th:nth-child(1) { width: 20%; }
    th:nth-child(2) { width: 15%; }
    th:nth-child(3) { width: 15%; }
    th:nth-child(4) { width: 25%; }
    th:nth-child(5) { width: 25%; }
    tr:hover { background: #f5f7ff; cursor: pointer; }
    .status-pill {
      padding: 4px 10px;
      border-radius: 999px;
      font-size: 11px;
      font-weight: 700;
      display: inline-flex;
      align-items: center;
      gap: 4px;
      text-transform: uppercase;
      letter-spacing: 0.03em;
      white-space: nowrap;
    }
    .status-passed { background: #e3f9e5; color: #037971; }
    .status-failed { background: #ffeff0; color: #c81e1e; }
    .status-skipped { background: #fff7e8; color: #ad6200; }
    .status-timedOut { background: #fff4e6; color: #b44d12; }
    .status-flaky { background: #edf2ff; color: #364fc7; }
    .pill {
      background: #edf2ff;
      border: 1px solid #c3d4ff;
      padding: 4px 8px;
      border-radius: 999px;
      font-size: 12px;
      color: #364fc7;
      display: inline-flex;
      gap: 4px;
      align-items: center;
    }
    .attachments { 
      display: flex; 
      gap: 8px; 
      flex-wrap: wrap;
      max-height: 36px;
      overflow: hidden;
      align-items: center;
    }
    .attachments a { 
      color: #3366ff; 
      text-decoration: none; 
      font-size: 12px;
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
    }
    .attachments a:hover { text-decoration: underline; }
    .attachments .muted {
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
    }
    .progress {
      position: relative;
      width: 100%;
      height: 10px;
      border-radius: 999px;
      overflow: hidden;
      background: #edf2ff;
      margin-top: 6px;
    }
    .progress-bar {
      position: absolute;
      inset: 0;
      width: var(--pct, 0%);
      background: linear-gradient(90deg, #4c6fff, #6ee7b7);
      transition: width 220ms ease-out;
    }
    .flex { display: flex; gap: 12px; align-items: center; flex-wrap: wrap; }
    .mono { font-family: ui-monospace, SFMono-Regular, Menlo, Consolas, monospace; }
    .scroll { overflow: auto; max-height: 320px; }
    .cell-content {
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
    }
    @media (max-width: 680px) { header { flex-direction: column; align-items: flex-start; gap: 10px; } }
  </style>
</head>
<body>
  <header>
    <div>
      <h1>Playwright Report Viewer</h1>
      <div class="muted">Custom JSON feed from custom-report/&lt;runId&gt;/results.json</div>
    </div>
    <div class="controls">
      <label for="runSelect" class="muted">Run:</label>
      <select id="runSelect"></select>
      <button id="refresh">Refresh</button>
    </div>
  </header>
  <main>
    <div class="grid" id="summaryGrid"></div>
    <div class="grid" style="margin-top:16px;">
      <div class="card">
        <h3>Tests</h3>
        <div class="scroll">
          <table id="testsTable">
            <thead>
              <tr>
                <th>Title</th>
                <th>Status</th>
                <th>Duration</th>
                <th>File</th>
                <th>Attachments</th>
              </tr>
            </thead>
            <tbody></tbody>
          </table>
        </div>
      </div>
      <div class="card" id="detailsCard">
        <h3>Details</h3>
        <div id="detailsContent" class="muted">Click a test row to see detailed steps, errors and analytics.</div>
      </div>
    </div>
  </main>
  <script>
    const runSelect = document.getElementById('runSelect');
    const refreshBtn = document.getElementById('refresh');
    const summaryGrid = document.getElementById('summaryGrid');
    const testsBody = document.querySelector('#testsTable tbody');
    let currentData = null;

    refreshBtn.addEventListener('click', loadRuns);
    runSelect.addEventListener('change', () => loadRun(runSelect.value));

    async function loadRuns() {
      const res = await fetch('/api/runs').then(r => r.json());
      runSelect.innerHTML = '';
      res.forEach((run, idx) => {
        const opt = document.createElement('option');
        opt.value = run.runId;
        opt.textContent = \`\${run.runId} (tests: \${run.tests}, fail: \${run.failed})\`;
        if (idx === 0) opt.selected = true;
        runSelect.appendChild(opt);
      });
      if (res.length) loadRun(res[0].runId);
    }

    async function loadRun(runId) {
      if (!runId) return;
      const data = await fetch('/api/run/' + runId).then(r => r.json());
      currentData = data;
      renderSummary(data);
      renderTests(data);
    }

    function renderSummary(data) {
      const t = data.totals;
      const durationMin = (t.durationMs / 60000).toFixed(2);
      summaryGrid.innerHTML = '';
      const cards = [
        { title: 'Overall', body: \`
          <div class="metric">\${t.passed}/\${t.tests} passed</div>
          <div class="progress" style="--pct:\${pct(t.passed,t.tests)}%">
            <div class="progress-bar"></div>
          </div>
          <div class="muted">duration \${durationMin} min</div>\`
        },
        { title: 'Statuses', body: \`
          <div class="flex">
            <span class="badge pass">Passed \${t.passed}</span>
            <span class="badge fail">Failed \${t.failed}</span>
            <span class="badge warn">Timed Out \${t.timedOut}</span>
            <span class="badge skip">Skipped \${t.skipped}</span>
            <span class="badge">Flaky \${t.flaky}</span>
          </div>\`
        },
        { title: 'Run info', body: \`
          <div class="muted">Run ID: <span class="mono">\${data.runId}</span></div>
          <div class="muted">Started: \${fmt(data.startedAt)}</div>
          <div class="muted">Finished: \${fmt(data.finishedAt)}</div>
          <div class="muted">Projects: \${(data.projects||[]).join(', ')}</div>\`
        }
      ];
      cards.forEach(c => {
        const card = document.createElement('div');
        card.className = 'card';
        card.innerHTML = \`<h3>\${c.title}</h3>\${c.body}\`;
        summaryGrid.appendChild(card);
      });
    }

    function renderTests(data) {
      testsBody.innerHTML = '';
      (data.tests || []).forEach(test => {
        const tr = document.createElement('tr');
        const attachments = (test.attachments || []).map(a => {
          if (a.path) {
            const href = '/attachment?p=' + encodeURIComponent(a.path);
            return \`<a href="\${href}" target="_blank" title="\${a.name}">\${a.name}</a>\`;
          }
          return \`<span class="muted" title="\${a.name}">\${a.name}</span>\`;
        }).join(' ');
        tr.innerHTML = \`
          <td><div class="cell-content" title="\${test.title}">\${test.title}</div></td>
          <td><span class="status-pill status-\${test.status}">\${test.status}</span></td>
          <td class="mono"><div class="cell-content">\${ms(test.durationMs)}</div></td>
          <td class="muted"><div class="cell-content" title="\${test.file || ''}">\${test.file || ''}</div></td>
          <td><div class="attachments">\${attachments || ''}</div></td>
        \`;
        tr.addEventListener('click', () => renderDetails(test));
        testsBody.appendChild(tr);
      });
    }

    function renderDetails(test) {
      const el = document.getElementById('detailsContent');
      if (!test) {
        el.textContent = 'Click a test row to see detailed steps, errors and analytics.';
        return;
      }
      const steps = (test.steps || []).map(step => \`
        <li><span class="mono">\${step.category || ''}</span> – \${step.title}
          <span class="muted">(\${ms(step.durationMs)})</span>
          \${step.error ? '<div class="muted">Error: ' + escapeHtml(step.error.message || '') + '</div>' : ''}
        </li>\`).join('');
      const attachments = (test.attachments || []).map(a => {
        if (a.path) {
          const href = '/attachment?p=' + encodeURIComponent(a.path);
          return \`<a href="\${href}" target="_blank">\${a.name}</a>\`;
        }
        return \`<span class="muted">\${a.name}</span>\`;
      }).join(' ');
      el.innerHTML = \`
        <div class="flex" style="margin-bottom:8px;">
          <span class="status-pill status-\${test.status}">\${test.status}</span>
          <span class="pill mono">\${ms(test.durationMs)}</span>
        </div>
        <div class="muted">File: <span class="mono">\${test.file || ''}</span></div>
        <div class="muted">Location: line \${test.line || '?'} col \${test.column || '?'}</div>
        <div style="margin-top:10px;">
          <strong>Attachments:</strong>
          <div class="attachments">\${attachments || '<span class="muted">None</span>'}</div>
        </div>
        <div style="margin-top:10px;">
          <strong>Steps:</strong>
          <ul>\${steps || '<li class="muted">No step data</li>'}</ul>
        </div>
        \${test.error ? '<div style="margin-top:10px;"><strong>Error:</strong><pre class="muted" style="white-space:pre-wrap;">' +
          escapeHtml(test.error.stack || test.error.message || '') + '</pre></div>' : ''}
      \`;
    }

    const pct = (part, total) => total ? Math.round((part / total) * 100) : 0;
    const ms = (n) => n != null ? n + ' ms' : '';
    const fmt = (d) => d ? new Date(d).toLocaleString() : '';
    const escapeHtml = (s) => s.replace(/[&<>\"']/g, c => ({'&':'&amp;','<':'&lt;','>':'&gt;','\"':'&quot;',\"'\":'&#39;'}[c]));

    loadRuns();
  </script>
</body>
</html>`;

  const sendJson = (res, status, data) => {
    res.writeHead(status, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify(data));
  };

  const listRuns = () => {
    if (!fs.existsSync(reportRoot)) return [];
    return fs
      .readdirSync(reportRoot, { withFileTypes: true })
      .filter((d) => d.isDirectory())
      .map((d) => d.name)
      .sort((a, b) => (a > b ? -1 : 1));
  };

  const loadRun = (runId) => {
    const file = path.join(reportRoot, runId, 'results.json');
    if (!fs.existsSync(file)) return null;
    return JSON.parse(fs.readFileSync(file, 'utf-8'));
  };

  const server = http.createServer((req, res) => {
    const parsed = url.parse(req.url, true);

    if (parsed.pathname === '/api/runs') {
      const runs = listRuns().map((id) => {
        const data = loadRun(id);
        return {
          runId: id,
          tests: data?.totals?.tests || 0,
          failed: data?.totals?.failed || 0,
          startedAt: data?.startedAt,
        };
      });
      return sendJson(res, 200, runs);
    }

    if (parsed.pathname && parsed.pathname.startsWith('/api/run/')) {
      const runId = parsed.pathname.split('/').pop();
      const data = loadRun(runId);
      if (!data) return sendJson(res, 404, { error: 'Run not found' });
      return sendJson(res, 200, data);
    }

    if (parsed.pathname === '/attachment') {
      const filePath = parsed.query && parsed.query.p;
      if (!filePath) {
        return sendJson(res, 400, { error: 'Missing path' });
      }
      const resolved = filePath;
      if (!fs.existsSync(resolved)) {
        return sendJson(res, 404, { error: 'Attachment not found' });
      }
      const stream = fs.createReadStream(resolved);
      res.writeHead(200);
      return stream.pipe(res);
    }

    // Serve viewer HTML
    res.writeHead(200, { 'Content-Type': 'text/html' });
    res.end(viewerHtml);
  });

  // Library returns server; caller is responsible for server.listen(...)
  server._port = port; // optional helper metadata
  return server;
}

module.exports = { createServer };
